// @astrojs/db/dist/core/errors.js
import { bold, cyan, red } from "kleur/colors";
const MISSING_SESSION_ID_ERROR = `${red("\u25B6 Login required!")}

  To authenticate with Astro Studio, run
  ${cyan("astro db login")}
`;
const MISSING_PROJECT_ID_ERROR = `${red("\u25B6 Directory not linked.")}

  To link this directory to an Astro Studio project, run
  ${cyan("astro db link")}
`;
const MISSING_EXECUTE_PATH_ERROR = `${red(
  "\u25B6 No file path provided."
)} Provide a path by running ${cyan("astro db execute <path>")}
`;
const RENAME_TABLE_ERROR = (oldTable, newTable) => {
  return red("\u25B6 Potential table rename detected: " + oldTable + ", " + newTable) + `
  You cannot add and remove tables in the same schema update batch.
  To resolve, add a 'deprecated: true' flag to '${oldTable}' instead.`;
};
const RENAME_COLUMN_ERROR = (oldSelector, newSelector) => {
  return red("\u25B6 Potential column rename detected: " + oldSelector + ", " + newSelector) + `
  You cannot add and remove columns in the same table.
  To resolve, add a 'deprecated: true' flag to '${oldSelector}' instead.`;
};
const FILE_NOT_FOUND_ERROR = (path) => `${red("\u25B6 File not found:")} ${bold(path)}
`;
const SHELL_QUERY_MISSING_ERROR = `${red(
  "\u25B6 Please provide a query to execute using the --query flag."
)}
`;
const SEED_ERROR = (error) => {
  return `${red(`Error while seeding database:`)}

${error}`;
};
const EXEC_ERROR = (error) => {
  return `${red(`Error while executing file:`)}

${error}`;
};
const SEED_DEFAULT_EXPORT_ERROR = (fileName) => {
  return SEED_ERROR(`Missing default function export in ${bold(fileName)}`);
};
const EXEC_DEFAULT_EXPORT_ERROR = (fileName) => {
  return EXEC_ERROR(`Missing default function export in ${bold(fileName)}`);
};
const REFERENCE_DNE_ERROR = (columnName) => {
  return `Column ${bold(
    columnName
  )} references a table that does not exist. Did you apply the referenced table to the \`tables\` object in your db config?`;
};
const FOREIGN_KEY_DNE_ERROR = (tableName) => {
  return `Table ${bold(
    tableName
  )} references a table that does not exist. Did you apply the referenced table to the \`tables\` object in your db config?`;
};
const FOREIGN_KEY_REFERENCES_LENGTH_ERROR = (tableName) => {
  return `Foreign key on ${bold(
    tableName
  )} is misconfigured. \`columns\` and \`references\` must be the same length.`;
};
const FOREIGN_KEY_REFERENCES_EMPTY_ERROR = (tableName) => {
  return `Foreign key on ${bold(
    tableName
  )} is misconfigured. \`references\` array cannot be empty.`;
};
const INTEGRATION_TABLE_CONFLICT_ERROR = (integrationName, tableName, isUserConflict) => {
  return red("\u25B6 Conflicting table name in integration " + bold(integrationName)) + isUserConflict ? `
  A user-defined table named ${bold(tableName)} already exists` : `
  Another integration already added a table named ${bold(tableName)}`;
};
export {
  EXEC_DEFAULT_EXPORT_ERROR,
  EXEC_ERROR,
  FILE_NOT_FOUND_ERROR,
  FOREIGN_KEY_DNE_ERROR,
  FOREIGN_KEY_REFERENCES_EMPTY_ERROR,
  FOREIGN_KEY_REFERENCES_LENGTH_ERROR,
  INTEGRATION_TABLE_CONFLICT_ERROR,
  MISSING_EXECUTE_PATH_ERROR,
  MISSING_PROJECT_ID_ERROR,
  MISSING_SESSION_ID_ERROR,
  REFERENCE_DNE_ERROR,
  RENAME_COLUMN_ERROR,
  RENAME_TABLE_ERROR,
  SEED_DEFAULT_EXPORT_ERROR,
  SEED_ERROR,
  SHELL_QUERY_MISSING_ERROR
};
